arcgis-ios-sdk-samples/Route and directions/Navigate route with rerouting/NavigateRouteWithReroutingViewController.swift

```diff
@@ -0,0 +1,364 @@
+    /// - Returns: An `AGSRouteTracker` object.
+    func makeRouteTracker(result: AGSRouteResult) -> AGSRouteTracker {
+        let tracker = AGSRouteTracker(routeResult: result, routeIndex: 0, skipCoincidentStops: true)!
+        tracker.delegate = self
+        tracker.voiceGuidanceUnitSystem = Locale.current.usesMetricSystem ? .metric : .imperial
+        return tracker
+    }
+    
+    /// Set route tracker, data source, and location display with a solved route result.
+    ///
+    /// - Parameter routeResult: An `AGSRouteResult` object.
+    func setNavigation(with routeResult: AGSRouteResult) {
+        // Set the route tracker
+        routeTracker = makeRouteTracker(result: routeResult)
+        
+        // Set the mock location data source.
+        let firstRoute = routeResult.routes.first!
+        directionsList = firstRoute.directionManeuvers
+        
+        let mockDataSource = makeDetourDataSource(polyline: makePolylineFromGPX())
+        initialLocation = mockDataSource.locations?.first
+        // Create a route tracker location data source to snap the location display to the route.
+        let routeTrackerLocationDataSource = AGSRouteTrackerLocationDataSource(routeTracker: routeTracker, locationDataSource: mockDataSource)
+        
+        // Set location display.
+        mapView.locationDisplay.dataSource = routeTrackerLocationDataSource
+        recenter()
+        
+        // Update graphics and viewpoint.
+        let firstRouteGeometry = firstRoute.routeGeometry!
+        updateRouteGraphics(remaining: firstRouteGeometry)
+        updateViewpoint(geometry: firstRouteGeometry)
+        
+        if routeTask.routeTaskInfo().supportsRerouting {
+            routeTracker.enableRerouting(with: routeTask, routeParameters: routeParameters!, strategy: .toNextWaypoint, visitFirstStopOnStart: false) { error in
+                if let error = error {
+                    self.presentAlert(error: error)
+                }
+            }
+        }
+    }
+    
+    /// Update the viewpoint so that it reflects the original viewpoint when the example is loaded.
+    ///
+    /// - Parameter result: An `AGSGeometry` object used to update the view point.
+    func updateViewpoint(geometry: AGSGeometry) {
+        // Show the resulting route on the map and save a reference to the route.
+        if let viewPoint = defaultViewPoint {
+            // Reset to initial view point with animation.
+            mapView.setViewpoint(viewPoint, completion: nil)
+        } else {
+            mapView.setViewpointGeometry(geometry) { [weak self] _ in
+                // Get the initial zoomed view point.
+                self?.defaultViewPoint = self?.mapView.currentViewpoint(with: .centerAndScale)
+            }
+        }
+    }
+    
+    // MARK: UI
+    
+    func setStatus(message: String) {
+        statusLabel.text = message
+    }
+    
+    // MARK: UIViewController
+    override func viewDidLoad() {
```

Comment:
Consider calling `reset()` on `viewWillDisappear`. 
Now when user go to README and src panes, the navigation is still ongoing and giving voice guidance, which is strange... 😅